core/src/main/java/org/mule/runtime/core/privileged/routing/CompositeRoutingException.java

```diff
@@ -54,41 +67,83 @@ public CompositeRoutingException(RoutingResult routingResult) {
-
-    for (Entry<String, Error> entry : routingResult.getFailures().entrySet()) {
-      String routeSubtitle = String.format("Route %s: ", entry.getKey());
-      MuleException muleException = ExceptionHelper.getRootMuleException(entry.getValue().getCause());
-      if (muleException != null) {
-        builder.append(routeSubtitle).append(muleException.getDetailedMessage());
-      } else {
-        builder.append(routeSubtitle)
-            .append("Caught exception in Exception Strategy: " + entry.getValue().getCause().getMessage());
+    Map<String, Pair<Error, EventProcessingException>> detailedFailures = getDetailedFailures();
+
+    if (detailedFailures.isEmpty()) {
+      builder.append(LEGACY_MESSAGE_TITLE).append(lineSeparator());
+      for (Entry<String, Error> entry : routingResult.getFailures().entrySet()) {
+        MuleException muleException = ExceptionHelper.getRootMuleException(entry.getValue().getCause());
+        Throwable exception = entry.getValue().getCause();
+        appendMessageForExceptions(builder, entry.getKey(), exception, muleException);
+      }
+    } else {
+      // provide information about the composite exception itself
+      builder.append(super.getDetailedMessage());
+      // get detailed information about exceptions that make up composite exception
+      builder.append(lineSeparator()).append(MESSAGE_SUB_TITLE).append(lineSeparator());
+      for (Entry<String, Pair<Error, EventProcessingException>> entry : detailedFailures.entrySet()) {
+        MuleException muleException = entry.getValue().getSecond();
+        Throwable exception = entry.getValue().getFirst().getCause();
+        appendMessageForExceptions(builder, entry.getKey(), exception, muleException);
       }
     }
+
     return builder.toString();
   }
 
+  private void appendMessageForExceptions(StringBuilder builder, String route, Throwable exception, MuleException muleException) {
+    String routeSubtitle = String.format("Route %s: ", route);
+    builder.append(lineSeparator());
+    if (muleException != null) {
+      builder.append(routeSubtitle).append(muleException.getDetailedMessage());
+    } else {
+      builder.append(routeSubtitle)
+          .append("Caught exception in Exception Strategy: " + exception.getMessage());
+    }
+  }
+
+  private Map<String, Pair<Error, EventProcessingException>> getDetailedFailures() {
+    Method getDetailedFailuresMethod = null;
+    Map<String, Pair<Error, EventProcessingException>> detailedFailures = emptyMap();
+    try {
+      getDetailedFailuresMethod = RoutingResult.class.getMethod("getFailuresWithExceptionInfo");
+      detailedFailures =
+          (Map<String, Pair<Error, EventProcessingException>>) getDetailedFailuresMethod.invoke(routingResult);
+    } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+      e.printStackTrace();
+      LOGGER.warn("Invalid Invocation, Expected method doesn't exist");
```

Comment:
Make this log a WARNING log instead of a DEBUG one. Include the method name so the message becomes meaningful.